src/core/server/ui_settings/ui_settings_client.ts

```diff
@@ -268,33 +282,56 @@ export class UiSettingsClient implements IUiSettingsClient {
     return filteredValues;
   }
 
-  /**
-   * group change into different scopes
-   * @param changes ui setting changes
-   * @returns [global, user]
-   */
-  private groupChanges(changes: Record<string, any>) {
-    const userLevelKeys = [] as string[];
-    Object.entries(this.defaults).forEach(([key, value]) => {
+  private groupSettingsKeys(defaults: NonNullable<UiSettingsServiceOptions['defaults']>) {
+    Object.entries(defaults).forEach(([key, value]) => {
+      if (Array.isArray(value.scope) && (value.scope?.length ?? 0) > 1) {
+        this.uiSettingsKeysWithMoreThanOneScope.push(key);
+      }
       if (
         value.scope === UiSettingScope.USER ||
         (Array.isArray(value.scope) && value.scope.includes(UiSettingScope.USER))
       ) {
-        userLevelKeys.push(key);
+        this.userLevelSettingsKeys.push(key);
+      }
+      if (
+        value.scope === UiSettingScope.WORKSPACE ||
+        (Array.isArray(value.scope) && value.scope.includes(UiSettingScope.WORKSPACE))
+      ) {
+        this.workspaceLevelSettingsKeys.push(key);
+      }
+      if (
+        value.scope === UiSettingScope.GLOBAL ||
+        (Array.isArray(value.scope) && value.scope.includes(UiSettingScope.GLOBAL))
+      ) {
+        this.globalLevelSettingsKeys.push(key);
       }
     });
+  }
+
+  /**
+   * group change into different scopes
+   * @param changes ui setting changes
+   * @returns [global, user]
+   */
+  private groupChanges(changes: Record<string, any>) {
     const userChanges = {} as Record<string, any>;
     const globalChanges = {} as Record<string, any>;
+    const workspaceChanges = {} as Record<string, any>;
 
     Object.entries(changes).forEach(([key, val]) => {
-      if (userLevelKeys.includes(key)) {
+      if (this.uiSettingsKeysWithMoreThanOneScope.includes(key)) {
+        // if this setting has more than one scope, we should not update this setting without specifying scope
+        throw new Error(`Unable to update "${key}", because it has multiple scopes`);
```

Comment:
Minor: do we have to create a dedicated `uiSettingsKeysWithMoreThanOneScope` for duplicate scope checking? Can we checking the scope directly when processing changes?

Reply:
Agree, check that from `this.defaults` should e good enough.